### PR TITLE
Update docker configuration

### PR DESCRIPTION
**Pull request by mfzha**
Streamline building process with a Dockerfile that doesn't rely on VSCode (maybe not necessary for our purposes)

Changes node and npm versions too to maintain consistency

### CODE REVIEW

**Comment by V13Axel** [comment]
First: Thanks! The package updates are great, and I appreciate the PR. I'll give the gulpfile changes and such a closer look when I get the chance.

Second: Can you explain your reasoning behind your changes to the `devcontainer.json` and **complete removal of** the .devcontainer Dockerfile? I agree with having a generalized Dockerfile, but breaking VSCode's devcontainer workflow to do it seems an odd choice to me.

**Comment by mfzha** [comment]
TLDR: we change everything because the original setup does not benefit us and serves to potentially complicate things.

Reasoning:

-  For our purposes it doesn't seem necessary to include VSCode's Dockerfile, since the only thing being done so far was installing `byobu`.
- This doesn't break VSCode's Containers extension workflow from my testing. Indeed, looking at [VSCodes's documentation for their dockerfile](https://github.com/microsoft/vscode-dev-containers/tree/main/containers/javascript-node), we would only gain things like `zsh`, `eslint`, `nvm` - these things should (philosophically) be installed explicitly in the Dockerfile if we need them (which we don't, not for an SPA of our size).
- Moreover, it's not immediately clear to a new dev (me) which version of node and `npm` are being used with this docker image: it could be any one of node 14.x, and any of npm 5.x, 6.x, or 7.x -- contrast this to using dockerfiles created by docker themselves.

Other changes in `devcontainer.json`: 

- setting `NODE_ENV` to development/production just makes it explicit which packages `npm` should be installing on build. 
- We move exposing port 8080 to the Dockerfile instead of leaving it in `devcontainer.json`, as part of making things more general and less tied to the VSCode environment.
- In not requiring a non-root user, I don't believe our use case requires it.

But some of this reasoning is more 'how I would do it', and not necessarily 'how it should be done' - I'm certainly not married to these changes if you guys would prefer to run things differently :)

Minor note: I realised I made an error in branching from the Gulp migration branch as opposed to from master. I won't do that again :(

**Comment by V13Axel** [comment]
The purpose of `.devcontainer` is to build a container designed for use as a full-fat environment for development (e.g. running the `vscode` remote dev server, letting you run `npm` commands and development server/watchers/etc. from within the container via VSCode's built-in terminal), treating it as your primary workspace rather than having to worry about the appropriate versions of packages installed on your host machine.

I personally installed it to our fork of KFC in order to avoid weird cross-project `node` version issues - And, more importantly, for getting my Linux-newbie teammate who uses WSL2 up and running quickly, without having to teach him how to manage/change Node versions or how to use `docker exec` in a generalized container. Of course, the version concerns were more important before the package updates (As anything newer than 10 was barfing on `npm install`, for me at least).

Moving the port forwarding into the generalized Dockerfile is fine... except your changes also:
* Removed the installation of `byobu` - This one's fair for a generalized Dockerfile, as no need for a terminal muxer if you're just building a deployment-ready container image.... But devcontainer is not that.
* Removed running the devcontainer as `node` (User 1000 in the MS node containers, matches most folks' local machines to avoid permissions issues in the bind-mounted development environment it's designed to be)
* Changed to a base `node` container. Again, great for deployment-ready. not for a devcontainer.
* Moved the port forwarding into the dockerfile ... Which again, is fine for a deployment container. But best-practices for devcontainer is to do it in the config json.

So ... Yeah, those changes do break it. I haven't gotten a chance to look too terribly closely at the rest yet, but for now I vote we put back the .devcontainer Dockerfile, revert the .devcontainer json changes, and have both Dockerfiles available for a given developer to choose their workflow from. :slightly_smiling_face:

**Comment by mfzha** [comment]
I see now the purpose and realise I was approaching things incorectly, thanks for making me aware of that.

On the other hand, I think having a consistent environment for all devs would reduce the need for troubleshooting later down the line.

So I propose this:

- Move `./Dockerfile` to `./devcontainer/Dockerfile`, and
- Changing it to install `byobu`, and
- Revert the changes `devcontainer.json` that exposed port 8080 and set default terminal to `/bin/bash`.

Why I'm not sure about reverting other things:

- I'm not as familiar with whether to run the container as `node` or as `root`, but wouldn't the 2nd have less headaches in the dev process, making it easier overall? 
- I still believe basing it off docker's dockerfiles as opposed to MS dockerfiles is the better way to go for troubleshooting problems that may arise from npm packages.

**Comment by V13Axel** [comment]
I disagree, and instead `.devcontainer` should be reverted wholesale and maintained separately from any other Docker assets - the devcontainer setup is optional, VSCode-specific, and built for the devs that want to run it using the VSCode devcontainer workflow based on MS's docker images. It is not designed to be a production environment, nor an environment for people who are not using VS Code, and should thus be entirely separate from any generalized Dockerfile that may apply to either of the other use-cases.

> I think having a consistent environment for all devs would reduce the need for troubleshooting...

I agree! If someone comes along who doesn't know enough about `node` to manage it and install it themselves, or has some special setup where they can't/won't/don't want to, then let's just tell them to use VSCode and run it using devcontainers, since that's what the maintaners do at the moment. If they choose to try and run the `npm` scripts some other way ... That's their fault. :wink: 

Jokes aside, if all you're doing is developing the app ... It's an SPA with no backend services. You shouldn't really **need** a whole Docker container just to serve it up to you as a developer, especially if the updated package versions work on the latest `node` and `npm` versions! That was the whole reason I created the devcontainer setup to begin with - Running `npm install` on my host machine resulted in some package or another complaining about node versions. So, I setup a devcontainer to have a quick way to get out of dependency hell while getting up and running, so we could get this fork off the ground quickly. 

And even if it _doesn't_ work with the latest node/npm, you could always just install the appropriate `node` version with `nvm`, and run `npm start` on your host or even just use an actual web server to serve it up for yourself. (I'm a fan of `Laravel Valet` for that - As I can just clone `some-repo` and go to `some-repo.test` in a browser immediately, in most cases. It's beautiful.)

---

**Instead**, if we're going to have a generalized docker setup, I want to have it for the right reasons: Pushing a docker image to DockerHub, and/or for setting up additional backend services that a dev might need to run ([Like we've done with Fantasy Calendar](https://github.com/fantasycalendar/Fantasy-Calendar) and its nginx/PHP-FPM/database/selenium/redis setup). Being able to bind-mount and also run commands inside it for development should just be a bonus on top of that. Which **still** means it should be entirely separate from the devcontainer setup.

As for your uncertainties:

* devcontainer attaches the whole repo as a bind-mount to the container, and then **runs vscode remote server** in that container, so all editing and command line interaction can be done within the container. Bind-mounts preserve file permissions and ownership, meaning creating new files or running any commands that create files (like npm install) inside the container running as `root` will result in files getting created as `root`, revoking permission to them to the user on the host machine. 
* I agree with this **for a generalized container**. Not for a VSCode .devcontainers environment.

At least for now, I'd like to avoid changing things about the development environment we're already using, since it works and we have higher priority things we'd like to handle. We will probably revisit the idea later.

**Comment by mfzha** [comment]
If the primary reason to use the MS devcontainer setup as-is (which isn't required by them - see [here](https://code.visualstudio.com/docs/remote/containers#_create-a-devcontainerjson-file)) was to get out of dependency hell on setup, and this same problem is solved when moving to Docker's own containers, what then is the benefit from using `mcr.microsoft.com/vscode/devcontainers/javascript-node:0-10` instead of `node:lts`?

Other than that, I can get onboard with everything else, but at the very least we should set `build.args.VARIANT=14`.

Adding `.dockerignore` to `.devcontainer` would also be a nice-to-have.

**Comment by V13Axel** [comment]
I agree with updating build.args.VARIANT to 14.

The MS devcontainers have ZSH (which I do use), along with various other packages and configuration changes that make them suitable for use as a dev environment. They run as user 1000, avoiding permissions issues, (Which, yes, I know you can do via docker-compose), along with having additional packages like zip, curl, and git (Mostly useful for showing current branch on the user's shell prompt)... Stuff that makes it nice to work with as a development environment.

In short: the answer to your question "what is the benefit?" is that **I like it**. It does exactly what it's supposed to the way it is right now, and I don't see a reason to go messing about with it unless you are targeting a specific problem that you have had with it that you feel other developers will need to also solve. 

_Unless that's the case_, leave the devcontainer setup alone besides the variant arg for now, and keep the new Dockerfile you've added for your own use as well. We will almost certainly revisit the topic later, when we have dealt with other, more pressing, matters.
